### PR TITLE
8255495: Support CDS Archived Heap for uncompressed oops

### PR DESCRIPTION
**Pull request by calvinccheung**
This proposed change adds support for uncompressed oops for CDS archived heap. It is needed
for supporting archived heap for ZGC in the future. For now, only G1GC is supported with this change.
During dump time with compressed oops disabled, the MaxHeapSize is set to 4GB. This is to ensure
the offset of any object in the heap from the bottom of the heap is always a 32-bit integer.

Passed CI tiers 1 - 4 and hs-tier7-rt tests.

Preliminary HelloWorld startup perf. improvement:

instr delta =    -68813432    -49.7140%
time  delta =      -14.720 ms -20.3599%

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255495](https://bugs.openjdk.java.net/browse/JDK-8255495): Support CDS Archived Heap for uncompressed oops

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7223/head:pull/7223` \
`$ git checkout pull/7223`

Update a local copy of the PR: \
`$ git checkout pull/7223` \
`$ git pull https://git.openjdk.java.net/jdk pull/7223/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7223`

View PR using the GUI difftool: \
`$ git pr show -t 7223`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7223.diff">https://git.openjdk.java.net/jdk/pull/7223.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/filemap.cpp [hunk 1]
```diff
@@ -2088,13 +2134,21 @@ void FileMapInfo::map_heap_regions_impl() {
...
     _heap_pointers_need_patching = true;
-    relocated_closed_heap_region_bottom = start_address_as_decoded_from_archive(si);
   }
   assert(is_aligned(relocated_closed_heap_region_bottom, HeapRegion::GrainBytes),
          "must be");
 
+  if (!UseCompressedOops && delta != 0) {
+    HeapShared::set_runtime_delta(delta);
+  }
+
```
Comment (iklam) [inline src/hotspot/share/cds/filemap.cpp:2151]: I think the code in the previous few blocks can be simplified by adding two utility functions:

```
// The address where the bottom of this shared heap region should be mapped
// at runtime
address FileMapInfo::heap_region_runtime_start_address(FileMapRegion* spc) {
  assert(UseSharedSpaces, "runtime only");
  spc->assert_is_heap_region();
  if (UseCompressedOops) {
    return start_address_as_decoded_from_archive(si);
  } else {
     header()->heap_begin() + si->mapping_offset() + HeapShared::set_runtime_delta();
  }
}

void FileMapInfo::set_shared_heap_runtime_delta(ptrdiff_t delta) {
  if (UseCompressedOops) {
    HeapShared::init_narrow_oop_decoding(narrow_oop_base() + delta, narrow_oop_shift());
  } else {
    HeapShared::set_runtime_delta(delta);
  }
}
```

Then, many of the `if (UseCompressedOops)` checks above can be eliminated.

File: src/hotspot/share/cds/filemap.cpp [hunk 2]
```diff
@@ -2148,7 +2202,13 @@ bool FileMapInfo::map_heap_regions(int first, int max,  bool is_open_archive,
...
     size_t size = si->used();
     if (size > 0) {
-      HeapWord* start = (HeapWord*)start_address_as_decoded_from_archive(si);
+      HeapWord* start;
+      if (UseCompressedOops) {
+        start = (HeapWord*)start_address_as_decoded_from_archive(si);
+      } else {
+        assert(is_aligned(si->mapping_offset(), sizeof(HeapWord)), "must be");
+        start = (HeapWord*)(header()->heap_begin() + si->mapping_offset() + HeapShared::runtime_delta());
+      }
```
Comment (iklam) [inline src/hotspot/share/cds/filemap.cpp:2211]: This block can also be replaced with

```
HeapWord* start = (HeapWord*)heap_region_runtime_start_address(si);
```
The `is_aligned(si->mapping_offset()` check can be moved into `heap_region_runtime_start_address()`

File: src/hotspot/share/cds/heapShared.hpp [hunk 3]
```diff
@@ -418,9 +421,21 @@ class HeapShared: AllStatic {
 
   // Run-time only
   static void clear_root(int index);
+
+  static void set_runtime_delta(ptrdiff_t offset) {
```
Comment (iklam) [inline src/hotspot/share/cds/heapShared.hpp:425]: For consistency, `offset` should be renamed to `delta`.

File: src/hotspot/share/runtime/arguments.cpp [hunk 4]
```diff
@@ -3449,6 +3449,11 @@ void Arguments::set_shared_spaces_flags_and_archive_paths() {
   if (DumpSharedSpaces || UseSharedSpaces) {
     init_shared_archive_paths();
   }
+
+  if (DumpSharedSpaces && !UseCompressedOops) {
+    FLAG_SET_CMDLINE(MaxHeapSize, (size_t)4 * G);
```
Comment (iklam) [inline src/hotspot/share/runtime/arguments.cpp:3454]: I think we should override MaxHeapSize only if it's greater than 4G. The user should be able to choose a smaller value.

File: test/hotspot/jtreg/runtime/cds/appcds/javaldr/HumongousDuringDump.java [hunk 5]
```diff
@@ -28,6 +28,7 @@
  * @library /test/lib /test/hotspot/jtreg/runtime/cds/appcds /test/hotspot/jtreg/runtime/cds/appcds/test-classes
  * @requires vm.cds.write.archived.java.heap
  * @requires vm.jvmti
+ * @requires vm.opt.final.UseCompressedOops == true
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/javaldr/HumongousDuringDump.java:31]: Why is this change needed?

**Review by iklam — COMMENTED** [review]
The changes look good overall. I have some suggestions for simplifying the code. Also a question on testing.

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/cds/filemap.cpp]: For `FileMapInfo::heap_region_runtime_start_address`, I have modified it slightly to the following because
the first time it is called `_runtime_delta` hasn't been set yet.

```
address FileMapInfo::heap_region_runtime_start_address(FileMapRegion* spc, ptrdiff_t delta) {
  assert(UseSharedSpaces, "runtime only");
  spc->assert_is_heap_region();
  if (UseCompressedOops) {
    return start_address_as_decoded_from_archive(spc);
  } else {
    assert(is_aligned(spc->mapping_offset(), sizeof(HeapWord)), "must be");
    return header()->heap_begin() + spc->mapping_offset() + delta;
  }
}
```

File: src/hotspot/share/cds/filemap.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/share/cds/filemap.cpp:2220]: Done.

File: src/hotspot/share/cds/heapShared.hpp (hunk 3 above)
Comment (calvinccheung) [reply src/hotspot/share/cds/heapShared.hpp]: Done.

File: src/hotspot/share/runtime/arguments.cpp (hunk 4 above)
Comment (calvinccheung) [reply src/hotspot/share/runtime/arguments.cpp]: Per our off list discussions, the InitialHeapSize, MinHeapSize, and MaxHeapSize are being set in
`universe_init()` via `MetaspaceShared::adjust_heap_sizes_for_dumping()`.

File: test/hotspot/jtreg/runtime/cds/appcds/javaldr/HumongousDuringDump.java (hunk 5 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/javaldr/HumongousDuringDump.java]: I ran the test with the -XX:-UseCompressedOops and without the change and it passed. I've reverted the change.

**Review by iklam — APPROVED** [review]
LGTM. The start-up performance improvement is good! Thanks.

**Review by tschatzl — APPROVED** [review]
GC changes look good to me.
